pkg/operator/controller/nodeobservability/service_test.go

```diff
@@ -20,73 +20,220 @@ import (
+			existingObjects: []runtime.Object{
+				testControllerService(
+					podName,
+					test.TestNamespace,
+					map[string]string{"app": "nodeobservability-old", "nodeobs_cr": "test"},
+					map[string]string{injectCertsKey: podName},
+				),
+			},
+			deployment: &appsv1.Deployment{
+				Spec: appsv1.DeploymentSpec{
+					Selector: &metav1.LabelSelector{
+						MatchLabels: map[string]string{"app": "nodeobservability", "nodeobs_cr": "test"},
+					},
+				},
+			},
+			expectedService: testControllerService(
+				podName,
+				test.TestNamespace,
+				map[string]string{"app": "nodeobservability", "nodeobs_cr": "test"},
+				map[string]string{injectCertsKey: podName},
+			),
+		},
+		{
+			name: "existing service, ports modified",
+			existingObjects: []runtime.Object{
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{Name: "node-observability-agent", Namespace: "test-namespace"},
+					Spec: corev1.ServiceSpec{
+						ClusterIP: corev1.ClusterIPNone,
+						Type:      corev1.ServiceTypeClusterIP,
+						Ports: []corev1.ServicePort{
+							{
+								Protocol:   corev1.ProtocolTCP,
+								Port:       9440,
+								TargetPort: intstr.FromInt(9440),
+							},
+						},
+						Selector: map[string]string{"app": "nodeobservability", "nodeobs_cr": "test"},
+					},
+				},
+			},
+			deployment: &appsv1.Deployment{
+				Spec: appsv1.DeploymentSpec{
+					Selector: &metav1.LabelSelector{
+						MatchLabels: map[string]string{"app": "nodeobservability", "nodeobs_cr": "test"},
+					},
+				},
+			},
+			expectedService: testControllerService(
+				podName,
+				test.TestNamespace,
+				map[string]string{"app": "nodeobservability", "nodeobs_cr": "test"},
+				map[string]string{injectCertsKey: podName},
+			),
 		},
 		{
-			name: "Exists",
+			name: "existing service, service type modified",
 			existingObjects: []runtime.Object{
-				makeService(),
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{Name: "node-observability-agent", Namespace: "test-namespace"},
+					Spec: corev1.ServiceSpec{
+						ClusterIP: corev1.ClusterIPNone,
+						Type:      corev1.ServiceTypeNodePort,
+						Ports: []corev1.ServicePort{
+							{
+								Protocol:   corev1.ProtocolTCP,
+								Port:       port,
+								TargetPort: intstr.FromInt(targetPort),
+							},
+						},
+						Selector: map[string]string{"app": "nodeobservability", "nodeobs_cr": "test"},
+					},
+				},
 			},
-			expectedExist: true,
-			expectedSA:    makeService(),
+			deployment: &appsv1.Deployment{
+				Spec: appsv1.DeploymentSpec{
+					Selector: &metav1.LabelSelector{
+						MatchLabels: map[string]string{"app": "controller"},
+					},
+				},
+			},
+			expectedService: testControllerService(
+				podName,
+				test.TestNamespace,
+				map[string]string{"app": "nodeobservability", "nodeobs_cr": "test"},
+				map[string]string{injectCertsKey: podName},
+			),
+		},
+		{
+			name: "existing service, extra annotations present",
```

Comment:
Would be good to highlight on the fact the the annotation is accepted/kept. Because just by looking at the name, it's not clear whether the ensure function is supposed to remove or keep it.